packages/did/src/did-error.ts

```diff
@@ -0,0 +1,51 @@
+export class DidError extends Error {
+  constructor(
+    public readonly did: string,
+    message: string,
+    public readonly code: string,
+    public readonly status = 400,
+    cause?: unknown,
+  ) {
+    super(message, { cause })
+  }
+
+  /**
+   * For compatibility with error handlers in common HTTP frameworks.
+   */
+  get statusCode() {
+    return this.status
+  }
+
+  override toString() {
+    return `${this.constructor.name} ${this.code} (${this.did}): ${this.message}`
+  }
+
+  static from(cause: unknown, did: string): DidError {
+    if (cause instanceof DidError) {
+      return cause
+    }
+
+    const message =
+      cause instanceof Error
+        ? cause.message
+        : typeof cause === 'string'
+          ? cause
+          : 'An unknown error occurred'
+
+    const status =
+      (typeof (cause as any)?.statusCode === 'number'
```

Comment:
Because it's `unknown`, I can't.

![Capture d’écran 2024-06-03 à 09 18 44](https://github.com/bluesky-social/atproto/assets/813661/ffec1851-d8eb-45ef-95b4-beecb4493bb6)


I could add more runtime type assertions checks but this is actually one of the few cases where `any` makes sense in typescript IMO. It's basically me saying "I know what I'm doing. And I know it won't cause a runtime issue".